### PR TITLE
Allow multiple bodies in let statements

### PR DESCRIPTION
**Pull request by lgalabru**
Address https://github.com/blockstack/blockstack-core/issues/1042

### CODE REVIEW

File: src/vm/docs/mod.rs [hunk 1]
```diff
@@ -267,13 +267,13 @@ which must return the same type. In the case that the boolean input is `true`, t
...
-    input_type: "((name2 AnyType) (name2 AnyType) ...), A",
+    input_type: "((name2 AnyType) (name2 AnyType) ...), AnyType, ... A",
     output_type: "A",
-    signature: "(let ((name1 expr1) (name2 expr2) ...) expr-body)",
+    signature: "(let ((name1 expr1) (name2 expr2) ...) expr-body1 expr-body2 ... expr-body-last)",
     description: "The `let` function accepts a list of `variable name` and `expression` pairs,
 evaluating each expression and _binding_ it to the corresponding variable name. The _context_
-created by this set of bindings is used for evaluating and return the value of `expr-body`.",
-    example: "(let ((a 2) (b (+ 5 6 7))) (+ a b)) ;; Returns 20"
+created by this set of bindings is used for evaluating its body expressions and return the value of the last such expression.",
```
Comment (kantai) [inline src/vm/docs/mod.rs]: Would use this as an opportunity to fix the existing typo:

> created by this set of bindings is used for evaluating its body expressions and return the value of the last such expression.",

Should be:

> created by this set of bindings is used for evaluating its body expressions. The `let` expression returns the value of the last such body expression.",

File: src/vm/functions/mod.rs [hunk 2]
```diff
@@ -343,8 +343,16 @@ fn special_let(args: &[SymbolicExpression], env: &mut Environment, context: &Loc
             inner_context.variables.insert(binding_name, binding_value);
         }
 
-        // evaluate the let-body
-        eval(&args[1], env, &inner_context)
+        // evaluate the let-bodies
+        let mut results = Vec::new();
```
Comment (kantai) [inline src/vm/functions/mod.rs]: Here, you'd want to escalate errors in the bodies before evaluating subsequent ones (and you can also avoid using a Vec by using an Option) e.g.:

```rust
let mut last_result = none;
for body in args[1..].iter() {
  body_result = eval(&body, env, &inner_context)?;
  last_result.replace(body_result);
}

last_result
  .ok_or_err(UncheckedError::InvalidArguments("Must pass at least 1 body to (let ...)".to_string()).into())
```

This way, if an arithmetic error or an `expects!` occurred in the first body the subsequent bodies _won't be processed_, which is probably what we want. We should add a test case for this behavior (and possibly a note in the documentation for `let`).

Comment (lgalabru) [reply src/vm/functions/mod.rs]: 🙏! 
Addessed with https://github.com/blockstack/blockstack-core/pull/1049/commits/b2bf90bb7cd0f18122d0e3f7d6aa7f07ca38ebe7

File: src/vm/docs/mod.rs (hunk 1 above)
Comment (lgalabru) [reply src/vm/docs/mod.rs]: Addessed with https://github.com/blockstack/blockstack-core/pull/1049/commits/b2bf90bb7cd0f18122d0e3f7d6aa7f07ca38ebe7

**Comment by jcnelson** [comment]
I wish I'd seen this issue earlier, so I could have commented on it instead of doing it here (after the code was already written).  What is wrong with simply using `(begin (...))` for situations like this?  Isn't that what it's for?  If we accept this, then why not just drop `(begin (...))` altogether and allow _any_ control statement to have multiple bodies?

**Comment by kantai** [comment]
> I wish I'd seen this issue earlier, so I could have commented on it instead of doing it here (after the code was already written). What is wrong with simply using (begin (...)) for situations like this? Isn't that what it's for? If we accept this, then why not just drop (begin (...)) altogether and allow any control statement to have multiple bodies?

This is a fairly standard syntactic nicety in LISPs -- `let` and `define` usually allow this syntax (other statements usually do not, because they have additional arguments, e.g., you couldn't do this with an `if`). It's common enough that I think it's surprising _not_ to support this syntax.

**Comment by jcnelson** [comment]
> This is a fairly standard syntactic nicety in LISPs -- let and define usually allow this syntax (other statements usually do not, because they have additional arguments, e.g., you couldn't do this with an if). It's common enough that I think it's surprising not to support this syntax.

Ah, okay -- this addresses my main concern, which was that this would lead to developer surprise.  If it is already expected that there can be two ways to express a `define` or `let` body (with our without the `(begin`), then this code looks good to me.

**Pull request merged**